draft-ietf-mls-protocol.md

```diff
@@ -2368,8 +2374,14 @@ If there are multiple proposals that apply to the same leaf, the committer
 chooses one and includes only that one in the Commit, considering the rest
 invalid. The committer MUST prefer any Remove received, or the most recent
 Update for the leaf if there are no Removes. If there are multiple Add proposals
-for the same client, the committer again chooses one to include and considers
-the rest invalid.
+referencing the same `key_package`, the committer again chooses one to include
+and considers the rest invalid.
+
+If there are multiple Add proposals of KeyPackages with the same identity or
+signature key, the committer MUST choose one and consider the rest invalid.
+Similarly, if there are Add proposals of KeyPackages with an identity or
+signature key that is already present as a member of the group, the committer
+MUST consider them invalid.
```

Comment:
I still object to this requirement, on the same grounds as before.  Namely, multi-device cases will result in multiple appearances of the same identity with high probability.

I thought we had agreed to make the prohibition on duplication of signature keys.  If that's the case, then let's make that change and merge.

Reply:
Is there a particular reason why you wouldn't want to use different identities for the devices? They will have to be distinguished at some point. Even if you use X509 certificates, you could still use the fingerprint as identity.

This change would allow us to address group members by their identity as opposed to their leaf index, which makes for a more intuitive API and allows us to hide the fact that we're using a tree underneath.

Reply:
Well, if you define "identity" in the right way, sure :)  

Part of the challenge here is that we don't really have a well-defined notion of "identity" on which to hang requirements such as these.  And I would argue that we should not have such a notion at this level -- it's up to the application to define how the credentials in MLS are used to authenticate application-level identities.  "Identity" is also a bad API concept, because you're going to have different cases; think "remove this specific device" vs. "remove all participants with this phone number".

Your X.509 example seems like it points back to my suggestion of focusing this on signature keys.  Aside from serial numbers, public keys are the major reason that two certificates would be different.

Reply:
I'm less concerned about concrete API schemes, and more about the fact that we can't abstract the tree away.

I see that signature keys would allow us to hide the tree under an abstraction. However, signature keys of a given client, even if they are unique, (potentially) change over the time of that client's group membership, so if we used that to address individual group members, we would have to have an additional mapping that keeps track of changing signature keys, which is not ideal.

(I realize now that that means that my fingerprint example wouldn't work, either.)

You are right in that we should probably talk a bit about what "identity" really means in the context of MLS. As you said, it's the job of the application layer to verify credentials, so what is verified by the AS could be something more than the `identity` field that the spec currently requires. Looking at it from that angle: does the `identity` field currently serve any purpose in the spec other than that it's verified as part of the credential and that it doesn't change through updates?

If that is not the case, we could require the identity field to be something unique, while leaving other fields in the credential (that MLS doesn't have to know about) as the "user-facing" identity, which (also) gets verified by the AS as part of the credential and which the user can base their for authentication decision on. Thus, for MLS, the `identity` field, would be mostly administrative.